theano/gpuarray/dnn.py

```diff
@@ -2732,6 +2732,176 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+    def c_code_cache_version(self):
+        return (super(GpuDnnSpatialTfDesc, self).c_code_cache_version(), version())
+
+
+class GpuDnnGridGenerator(DnnBase):
+
+    """
+    This Op builds a spatial transformer grid generator for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('dtype',)
+    _cop_num_inputs = 3
+    _cop_num_outputs = 1
+
+    def __init__(self, dtype):
+        DnnBase.__init__(self, ["c_code/spatialtf_grid.c"], "spatialtf_grid")
+        self.dtype = dtype
+
+    def make_node(self, grid_dimensions, theta, desc):
+        context_name = infer_context_name(desc, theta)
+
+        grid_dimensions = as_tensor_variable(grid_dimensions)
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+
+        assert theta.dtype in ('float16', 'float32', 'float64')
+        assert cudnn.cudnnDataType_t.has_alias(theta.dtype)
+
+        # Allocate GPU memory for grid of coordinates
+        grid = GpuArrayType(dtype=self.dtype,
+                            broadcastable=(False, False, False, False,),
+                            context_name=context_name)()
+
+        return Apply(self, [grid_dimensions, theta, desc], [grid])
+
+    def L_op(self, inputs, outputs, output_grads):
+        pass
+
+
+class GpuDnnGridSampler(DnnBase):
+
+    """
+    This Op builds a spatial transformer grid sampler for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('dtype',)
+    _cop_num_inputs = 5
+    _cop_num_outputs = 1
+
+    def __init__(self, dtype):
+        DnnBase.__init__(self, ["c_code/spatialtf_sampler.c"], "spatialtf_sampler")
+        self.dtype = dtype
+
+    def make_node(self, img, grid, desc, alpha=None, beta=None):
+        context_name = infer_context_name(img, grid)
+
+        img = as_gpuarray_variable(img, context_name)
+        grid = as_gpuarray_variable(grid, context_name)
+
+        output = GpuArrayType(dtype=self.dtype,
+                              broadcastable=img.type.ndim * (False,),
+                              context_name=context_name)()
+
+        if img.type.ndim != 4:
+            raise TypeError('img must be a 4D tensor')
+
+        if (not isinstance(desc.type, CDataType) or
+                desc.type.ctype != 'cudnnSpatialTransformerDescriptor_t'):
+            raise ValueError('desc must be cudnnSpatialTransformerDescriptor_t')
+
+        alpha = ensure_dt(alpha, _one, 'alpha', img.dtype)
+        beta = ensure_dt(beta, _zero, 'beta', img.dtype)
+
+        return Apply(self, [img, grid, desc, alpha, beta], [output])
+
+    def L_op(self, inputs, outputs, output_grads):
+        pass
+
+
+def dnn_spatialtf(img, theta, grid_dims, alpha=None, beta=None, dtype=None):
+    """
+        GPU spatial transformer using cuDNN from NVIDIA.
+    """
+
+    # img is a 4D tensor with shape: (num_images, num_channels, width, height)
+    assert img.ndim == 4
+    # Grid dimensions must be a 4-dimensional tuple
```

Comment:
"a 4-length tuple" instead, I think. Also, I think you should allow iterable objects (tuples, or lists).

```python
# instead of `assert isinstance(grid_dims, tuple)`:
grid_dims = tuple(grid_dims)
```